src/lua/lua_thread_pool.c

```diff
@@ -0,0 +1,125 @@
+#include "config.h"
+
+#include <assert.h>
+
+#include "lua_common.h"
+#include "lua_thread_pool.h"
+
+struct lua_thread_pool {
+	GQueue *available_items;
+	lua_State *L;
+	gint max_items;
+	struct thread_entry *running_entry;
+};
+
+static struct thread_entry *
+thread_entry_new (lua_State * L)
+{
+	struct thread_entry *ent;
+	ent = g_malloc (sizeof *ent);
+	ent->lua_state = lua_newthread (L);
+	ent->thread_index = luaL_ref (L, LUA_REGISTRYINDEX);
+	return ent;
+}
+
+static void
+thread_entry_free (lua_State * L, struct thread_entry *ent)
+{
+	luaL_unref (L, LUA_REGISTRYINDEX, ent->thread_index);
+	g_free (ent);
+}
+
+struct lua_thread_pool *
+lua_thread_pool_new (lua_State * L)
+{
+	struct lua_thread_pool * pool = g_new0 (struct lua_thread_pool, 1);
+
+	pool->L = L;
+	pool->max_items = 100;
+
+	pool->available_items = g_queue_new ();
+	int i;
+
+	struct thread_entry *ent;
+	for (i = 0; i < pool->max_items; i ++) {
+		ent = thread_entry_new (pool->L);
+		g_queue_push_head (pool->available_items, ent);
+	}
+
+	return pool;
+}
+
+void
+lua_thread_pool_free (struct lua_thread_pool *pool)
+{
+	struct thread_entry *ent = NULL;
+	while (!g_queue_is_empty (pool->available_items)) {
+		ent = g_queue_pop_head (pool->available_items);
+		thread_entry_free (pool->L, ent);
+	}
+	g_queue_free (pool->available_items);
+	g_free (pool);
+}
+
+struct thread_entry *
+lua_thread_pool_get(struct lua_thread_pool *pool)
+{
+	gpointer cur;
+	struct thread_entry *ent = NULL;
+
+	cur = g_queue_pop_head (pool->available_items);
+
+	if (cur) {
+		ent = cur;
+	}
+	else {
+		ent = thread_entry_new (pool->L);
```

Comment:
In this case thread is created but not pushed to the queue. It looks strange.

Reply:
The thread will be pushed into the queue once it's freed. The queue contains only available threads, while this functions makes thread unavailable.